### PR TITLE
WIP: Add basic deoplete hooks.

### LINKED ISSUE #460: Implement deoplete api
It would be super if clang_complete would interface with deoplete like, for example, neco-ghc.

It just seems to require a python file like this https://github.com/eagletmt/neco-ghc/blob/master/rplugin/python3/deoplete/sources/ghc.py implementing gather_candidates.

### PR DESCRIPTION
**Pull request by expipiplus1**
Closes #460 

I've not written any Python for a long time, and don't really know anything about the internals of deoplete or clang_complete however this seems to work very well in my limited testing. I'm sure that @Shougo has some useful things to say about this.

### CODE REVIEW

**Comment by xaizek** [comment]
What's `rplugin/` directory and why sub-directory is saying `python3`? I thought that clang_complete doesn't work with python3 (but there were necessary changes in one of forks).

Also, deoplete seems to be something for neovim, does this mean that clang_complete works there? (Just checking, because those guys like to break things and I wouldn't be surprised if the plugin doesn't work there.)

**Comment by expipiplus1** [comment]
`rplugin` stands for remote plugin I think. there's some documentation here: https://neovim.io/doc/user/remote_plugin.html

This is kind of like a meta-plugin for deoplete.

Although this is written in python3, it interfaces with the python2 code via vimscript :/ 
I'm sure it would be a little bit nicer if clang_complete was written in python3 too, and could be called directly.

Yes, this is for Neovim. Pretty much every plugin I use has worked with Neovim out of the box, clang_complete is no exception. This just makes the interface with deoplete a bit nicer, and consistent with other plugins.

**Comment by xaizek** [comment]
Thanks for the explanation, if the plugin is run in a separate process, it explains why it would work.

Seems harmless yet might be useful for someone, so I'm merging this, thanks.

**Comment by expipiplus1** [comment]
You're welcome.

I think it would be best to have someone who understands the deoplete side of things to take a look at this at some point.

File: rplugin/python3/deoplete/sources/clang_complete.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#=============================================================================
+# FILE: clang_complete.py
+# AUTHOR: Joe Hermaszewski
+#=============================================================================
+
+from .base import Base
+import deoplete.util
+import re
```
Comment (Shougo) [inline rplugin/python3/deoplete/sources/clang_complete.py]: `import re` and `import deoplete.util` are not needed.

File: rplugin/python3/deoplete/sources/clang_complete.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        Base.__init__(self, vim)
+
+        self.name = 'clang_complete'
+        self.mark = '[clang]'
+        self.filetypes = ['c', 'cpp']
+        self.is_bytepos = True
+        self.min_pattern_length = 1
+
+    def get_complete_position(self, context):
+        return self.vim.eval("ClangComplete(1, 0)")
```
Comment (Shougo) [inline rplugin/python3/deoplete/sources/clang_complete.py]: You should use `self.vim.call()`like this:

```python
return self.vim.call('ClangComplete', 1, 0)
```

File: rplugin/python3/deoplete/sources/clang_complete.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    def gather_candidates(self, context):
+        return self.vim.eval("ClangComplete(0, \"\")")
```
Comment (Shougo) [inline rplugin/python3/deoplete/sources/clang_complete.py]:
```python
return self.vim.call('ClangComplete', 0, '')
```

**Comment by Shougo** [comment]
Thanks.  I have reviewed it.

**Comment by expipiplus1** [comment]
Thanks!

**Pull request merged**